test_conformance/subgroups/test_workgroup_ballot.cpp

```diff
@@ -0,0 +1,973 @@
+                // Check result
+
+                for (i = 0; i < n; ++i)
+                { // for each subgroup
+                    int bit_value = 0;
+                    int bit_mask = 1
+                        << ((BallotOp::inverse_ballot == operation)
+                                ? i
+                                : l % 32); // from which value of bitfield bit
+                                           // verification will be done
+
+                    if (i < 32)
+                        (mx[ii + i].s0 & bit_mask) > 0 ? bit_value = 1
+                                                       : bit_value = 0;
+                    if (i >= 32 && i < 64)
+                        (mx[ii + i].s1 & bit_mask) > 0 ? bit_value = 1
+                                                       : bit_value = 0;
+                    if (i >= 64 && i < 96)
+                        (mx[ii + i].s2 & bit_mask) > 0 ? bit_value = 1
+                                                       : bit_value = 0;
+                    if (i >= 96 && i < 128)
+                        (mx[ii + i].s3 & bit_mask) > 0 ? bit_value = 1
+                                                       : bit_value = 0;
+
+                    bit_value == 1 ? tr = { 1, 0, 0, 0 } : tr = { 0, 0, 0, 0 };
+
+                    rr = my[ii + i];
+                    if (!compare(rr, tr))
+                    {
+                        log_error(
+                            "ERROR: sub_group_%s mismatch for local id %d in "
+                            "sub group %d in group %d obtained {%d, %d, %d, "
+                            "%d}, expected {%d, %d, %d, %d}\n",
+                            operation_names(operation), i, j, k, rr.s0, rr.s1,
+                            rr.s2, rr.s3, tr.s0, tr.s1, tr.s2, tr.s3);
+                        return -1;
+                    }
+                }
+            }
+            x += nw;
+            y += nw;
+            m += 4 * nw;
+        }
+
+        return 0;
+    }
+};
+
+
+// Test for bit count/inclusive and exclusive scan/ find lsb msb ballot function
+template <typename Ty, BallotOp operation> struct BALLOT3
+{
+    static void gen(Ty *x, Ty *t, cl_int *m, int ns, int nw, int ng)
+    {
+        int i, ii, j, k, n;
+        int nj = (nw + ns - 1) / ns;
+        int e;
+        ii = 0;
+        for (k = 0; k < ng; ++k)
+        { // for each work_group
+            for (j = 0; j < nj; ++j)
+            { // for each subgroup
+                ii = j * ns;
+                n = ii + ns > nw ? nw - ii : ns;
+                if (operation == BallotOp::ballot_bit_count
+                    || operation == BallotOp::ballot_inclusive_scan
+                    || operation == BallotOp::ballot_exclusive_scan)
+                {
+                    // Initialize data matrix indexed by local id and sub group
+                    // id
+                    e = (int)(genrand_int32(gMTdata) % 3);
+                    switch (e)
```

Comment:
This switch looks very similar to some I've seen above. It might be worth DRY-ing them and introduce a function.

Reply:
removed

Reply:
Fixed https://github.com/KhronosGroup/OpenCL-CTS/pull/1074/commits/0bf22751ded07d81b6ad948105046cac61a45476